### PR TITLE
comms: Add sol-netctl agent API V6

### PR DESCRIPTION
**Pull request by wzhen12**
The agent is necessary in the process of network connection,
that will receive input information from the user and send it
to the network(such as AP and so on).

The API is based on connman D-Bus APIs, it provides register agent,
input login information and so on netctl agent API.

Signed-off-by: Wu Zheng wu.zheng@intel.com

### CODE REVIEW

**Comment by wzhen12** [comment]
The Version 6 of Add sol-netctl agent API.
1. Modify the version 5 errors. (https://github.com/solettaproject/soletta/pull/2312)
2. Change the vector param of sol_netctl_request_input and request_input to ptr_vector. 

@ceolin and @barbieri Please check it.

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 1]
```diff
@@ -1212,3 +1241,380 @@ sol_netctl_service_state_to_str(enum sol_netctl_service_state state)
...
+    if (!_ctx.agent)
+        return -ENOENT;
+
+    _release_agent(&_ctx);
+
+    r = sd_bus_call_method_async(bus, NULL,
+        "net.connman", "/", "net.connman.Manager", "UnregisterAgent",
+        sol_bus_log_callback, NULL, "o", CONNMAN_AGENT_PATH);
+
+    SOL_INT_CHECK(r, < 0, r);
```
Comment (ceolin) [inline src/lib/comms/sol-netctl-impl-connman.c]: don't need this check, you can simply return `r`

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 2]
```diff
@@ -1618,3 +1623,45 @@ sol_netctl_unregister_agent(void)
...
+        "Scan", _scan_return, NULL, NULL);
+}
+
+SOL_API int
+sol_netctl_scan(void)
+{
+    int r;
+
+    r = _scan_services();
+    SOL_INT_CHECK(r, < 0, r);
```
Comment (ceolin) [inline src/lib/comms/sol-netctl-impl-connman.c]: `return _scan_services()`

File: src/lib/comms/sol-netctl-impl-connman.c [hunk 3]
```diff
@@ -1212,3 +1241,380 @@ sol_netctl_service_state_to_str(enum sol_netctl_service_state state)
...
+
+    r = sd_bus_message_close_container(reply);
+    SOL_INT_CHECK_GOTO(r, < 0, fail);
+
+    r = sd_bus_send(NULL, reply, NULL);
+    sd_bus_message_unref(reply);
+
+    _ctx.agent_msg = sd_bus_message_unref(_ctx.agent_msg);
+
+    return 0;
```
Comment (ceolin) [inline src/lib/comms/sol-netctl-impl-connman.c]: shouldn't return `r` here ? otherwise we will miss whether `sd_bus_send()` worked or not

**Comment by ceolin** [comment]
Please, review the comments and submit a new pr that I will integrate it. The serie is pretty good.

@wzhen12 is missing versioning netctl API, the structs in this module are not versioned. You don't need to do this for this pull request, since the problem exists before your pr.
Please let me know if you can do this in another pr.

**Comment by wzhen12** [comment]
@ceolin It is okay. Sorry, I have a Chinese holiday. I will have a response soon.

File: src/lib/comms/sol-netctl-impl-connman.c (hunk 1 above)
Comment (wzhen12) [inline src/lib/comms/sol-netctl-impl-connman.c]: okay

File: src/lib/comms/sol-netctl-impl-connman.c (hunk 2 above)
Comment (wzhen12) [inline src/lib/comms/sol-netctl-impl-connman.c]: okay

File: src/lib/comms/sol-netctl-impl-connman.c (hunk 3 above)
Comment (wzhen12) [inline src/lib/comms/sol-netctl-impl-connman.c]: okay

**Pull request closed without merging**